etherman/etherman.go

```diff
@@ -510,77 +572,122 @@ func (etherMan *Client) newBlocksEvent(ctx context.Context, vLog types.Log, bloc
 	return nil
 }
 
+func (etherMan *Client) l1BlockFromL2Block(ctx context.Context, l2Block SequencerBlock) (*types.Block, error) {
+	// Each L2 block must be associated with a unique L1 block, so that we know which global exit
+	// root (maintained on L1) to use when executing bridge withdrawals on L2. In the original
+	// Polygon zkEVM, this association is determined whenever an L2 batch is sequenced on L1. This
+	// design makes it impossible to use the HotShot sequencer for fast preconfirmations, because
+	// even though a canonical ordering of L2 blocks is determined quickly, we cannot execute those
+	// blocks until they have been persisted on L1, which can be slow.
+	//
+	// To enable fast preconfirmations, we redefine the way in which L2 blocks get associated with
+	// L1 blocks. Each time an L2 block is _sequenced_, the HotShot consensus protocol assigns it an
```

Comment:
I may be missing something, but how exactly does HotShot assign the L1 block number?

Reply:
It doesn't, this comment is aspirational. There's a big comment in the zkevm-adaptor query service in the other repo about how for the short term we assign the block numbers in the adaptor, not in consensus.